packages/worker/src/api/controllers/global/self.ts

```diff
@@ -105,16 +110,79 @@ export async function getSelf(ctx: any) {
   addSessionAttributesToUser(ctx)
 }
 
+export const syncAppFavourites = async (processedAppIds: string[]) => {
+  const apps = await fetchAppsByIds(processedAppIds)
+  return apps?.reduce((acc: string[], app) => {
+    const id = app.appId.replace(dbCore.APP_DEV_PREFIX, "")
+    if (processedAppIds.includes(id)) {
+      acc.push(id)
+    }
+    return acc
+  }, [])
+}
+
+export const fetchAppsByIds = async (processedAppIds: string[]) => {
+  return await dbCore.getAppsByIDs(
+    processedAppIds.map(appId => `${dbCore.APP_DEV_PREFIX}${appId}`)
+  )
+}
+
+const processUserAppFavourites = (
+  user: User,
+  appIdRequest: string[] | undefined
+) => {
+  const userAppFavourites = user.appFavourites || []
+  const appFavouritesRequest = appIdRequest || []
+  const appFavouritesUpdated = [
+    ...new Set(userAppFavourites.concat(appFavouritesRequest)),
+  ]
+
+  // Process state
+  let processedAppIds: string[]
+  if (userAppFavourites.length) {
+    const req = new Set(appFavouritesRequest)
+    const updated = new Set(userAppFavourites)
+    for (const element of req) {
+      if (updated.has(element)) {
+        // Subtract/Toggle any existing ones
+        updated.delete(element)
+      } else {
+        // Add new appIds
+        updated.add(element)
+      }
+    }
+    processedAppIds = [...updated]
+  } else {
+    processedAppIds = [...appFavouritesUpdated]
+  }
+  return processedAppIds
+}
+
 export async function updateSelf(
   ctx: UserCtx<UpdateSelfRequest, UpdateSelfResponse>
 ) {
   const update = ctx.request.body
 
   let user = await userSdk.db.getUser(ctx.user._id!)
-  user = {
-    ...user,
-    ...update,
+
+  if ("appFavourites" in update) {
```

Comment:
Not sure if this should be here. Is there a reason why is this not in a different endpoint that handles only this data?

Reply:
I didn't separate this into it's own endpoint because it follows the same pattern we have elsewhere for updating distinct user fields. The block that you linked should only ever be hit when explicitly modifying the `appFavourites` as it makes a call to pull apps and sync with the server. This should only ever be run when an update to self includes `appFavourites`.

I did make a change however to ensure that the behaviour on the endpoint is exactly the same when updating favourites as it current excludes anything else.

Reply:
Where else do we use this pattern? It does feel quite strange.

Reply:
I was just referring to the use of `updateSelf` to update individual fields on the user. The only deviation `appFavourites` has here is an optional sync to clear out invalid apps. I don't think this warrants a separate endpoint to handle. The flow here has been updated further after feedback from @aptkingston.